common/dynamicconfig/memory_client.go

```diff
@@ -63,19 +71,43 @@ func (d *MemoryClient) GetValue(key Key) []ConstrainedValue {
 	return nil
 }
 
+func (d *MemoryClient) Subscribe(f ClientUpdateFunc) (cancel func()) {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	d.subscriptionIdx++
+	id := d.subscriptionIdx
+	d.subscriptions[id] = f
+
+	return func() {
+		d.lock.Lock()
+		defer d.lock.Unlock()
+		delete(d.subscriptions, id)
+	}
+}
+
 func (d *MemoryClient) OverrideSetting(setting GenericSetting, value any) (cleanup func()) {
 	return d.OverrideValue(setting.Key(), value)
 }
 
 func (d *MemoryClient) OverrideValue(key Key, value any) (cleanup func()) {
 	d.lock.Lock()
-	defer d.lock.Unlock()
 
 	var idx atomic.Int64
 	idx.Store(int64(len(d.overrides)))
 
 	d.overrides = append(d.overrides, kvpair{valid: true, key: key, value: value})
 
+	subscriptions := expmaps.Values(d.subscriptions)
+	newValue := d.getValueLocked(key)
+
+	d.lock.Unlock()
+
+	changed := map[Key][]ConstrainedValue{key: newValue}
+	for _, update := range subscriptions {
+		update(changed)
```

Comment:
Technically you may get called back after you unsubscribe but that's fine.

Reply:
Hmm, that's a good point. The same is true of the file-based client (which I just copied this from).

Maybe it should just hold the lock for the duration? That lock is only taken when Collections are Started/Stopped, so that seems fine.

Reply:
Another approach is to have an atomic reference for each callback that gets replaced with nil when you unsubscribe

Reply:
Yeah, I don't think it's worth that complexity though

Reply:
Agree for the memory client but maybe for the file based one we would want that just to prevent holding the lock too long if some subscription callback is taking longer than expected.

Reply:
The only thing blocked is Collection Start/Stop, and that only happens at process startup/shutdown